ListValue.equals(ArrayValue)

SaschaPeukert opened the pull request:
This includes tests and an implementation for an equality check between ListValues and ArrayValues if they contain the same values (with the same ordering).
Introduced SequenceValue as an interface that is implemented by ListValue and ArrayValue. Actual equality check is statically done in AnyValues.

SaschaPeukert committed:
ListValue.equals(ArrayValue)

This includes tests and an implementation for an equality check between
ListValues and ArrayValues if they contain the same values.
Introduced SequenceValue as an interface that is implemented by
ListValue and ArrayValue. Actual equality check is statically done in
AnyValues.

SaschaPeukert committed:
Making CheckStyle happy

MishaDemianenko commented on community/values/src/main/java/org/neo4j/values/AnyValues.java:
```diff
@@ -32,4 +32,23 @@
+        {
+            AnyValue myValue = a.value( i );
+            AnyValue otherValue = b.value( i );
+            if ( myValue.equals( otherValue ) == false )
```
this line should be simplified to `!myValue.equals( otherValue )`

MishaDemianenko commented on community/values/src/main/java/org/neo4j/values/AnyValues.java:
```diff
@@ -32,4 +32,23 @@
     public static final Comparator<AnyValue> COMPARATOR =
             new AnyValueComparator( Values.COMPARATOR, VirtualValueGroup::compareTo );
+
+    public static boolean equalityOfValuesInSequences( SequenceValue a, SequenceValue b )
```
It looks like this method is responsible for equality check for all possible sequence and it's quite suites as the default implementation of aquals in SequenceValue instead.

MishaDemianenko commented on community/values/src/main/java/org/neo4j/values/SequenceValue.java:
```diff
@@ -0,0 +1,29 @@
+ */
+package org.neo4j.values;
+
+public interface SequenceValue
```
Missing java doc.

MishaDemianenko commented on community/values/src/main/java/org/neo4j/values/VirtualValue.java:
```diff
@@ -33,6 +33,10 @@
     @Override
     public final boolean equals( Object other )
     {
+        if ( (other instanceof SequenceValue) && (this instanceof SequenceValue) )
```
the whole goal of new values API was to remove casts as much as possible from all the equality methods, this change does not follow that paradigm (this place and all other updated equals methods) is there is any way we can avoid that?

MishaDemianenko commented on community/values/src/main/java/org/neo4j/values/virtual/ListValue.java:
```diff
@@ -118,11 +117,18 @@ public String toString()
     }
 
-    public int size()
+    public int length()
```
Missing`@Override` annotation here

MishaDemianenko reviewed (changes requested):
Couple of small comments and questions, also since I'm mostly on the side of this work would really like if @fickludd would take a look at proposed changes as soon as will have time

SaschaPeukert committed:
Included Feedback from @MishaDemianenko

Included changes to reflect all feedback from Mischa's review.
I removed all instanceOf that I previously introduced. Instead, it
will now check a method isSequence() that is introduced to AnyValue

MishaDemianenko commented on community/values/src/main/java/org/neo4j/values/VirtualValue.java:
```diff
@@ -29,15 +29,28 @@
+        // suppose that we will only ever call equals with Objects of AnyValue but just in case
+        try
+        {
+            if ( ((AnyValue) other).isSequenceValue() && this.isSequenceValue() )
```
doing
```
try
{
(AnyValue) other)
}
catch ( ClassCastException e )
{
return false
}
```
is much more slower and not that easy to read compared to using instanceOf operator, so would suggest here to use instanceOf AnyValue instead. 
 This also applies to ArrayValue `equals`

MishaDemianenko reviewed (commented):
One comment regarding casting and catching class cast otherwise looks good!

SaschaPeukert committed:
More feedback from @MishaDemianenko and tests

Reintroduced one instaceOf to get rid of the try-catch and added some
tests for mixed lists.

SaschaPeukert committed:
Moved equals into children of ArrayValue

I did not particularly like this change, but there seems to be no good
way that would make the checkstyle rule 'EqualsHashCode' happy
("all classes that overwrite one of those methods, must also overwrite
the other" -> but there is no good way to overwrite hashCode in
ArrayValue because inheritance)
